docs/DesignDocs/ExtendedEnclaveInitializationData.md

```diff
@@ -0,0 +1,154 @@
+
+First, before an enclave image with EEID is loaded, we patch the size settings from the signed image with the instance-specific settings.
+This actually alters the computation of the `MRENCLAVE` as the sequence of calls to `EADD/EEXTEND` is modified compared to the values used during signing.
+```C
+    if (eeid)
+    {
+        if (props.header.size_settings.num_heap_pages != 0 ||
+            props.header.size_settings.num_stack_pages != 0 ||
+            props.header.size_settings.num_tcs != 0)
+            OE_RAISE(OE_INVALID_PARAMETER);
+
+        props.header.size_settings.num_heap_pages =
+            eeid->size_settings.num_heap_pages;
+        props.header.size_settings.num_stack_pages =
+            eeid->size_settings.num_stack_pages;
+        props.header.size_settings.num_tcs = eeid->size_settings.num_tcs;
+
+        // patch
+        elf64_sym_t sym_props = {0};
+        const elf64_t* eimg = &oeimage.u.elf.elf;
+        if (elf64_find_symbol_by_name(
+                eimg, "oe_enclave_properties_sgx", &sym_props) == 0)
+        {
+            uint64_t* sym_props_addr = NULL;
+            sym_props_addr =
+                (uint64_t*)(oeimage.image_base + sym_props.st_value);
+            oe_sgx_enclave_properties_t* p =
+                (oe_sgx_enclave_properties_t*)sym_props_addr;
+            p->header.size_settings = props.header.size_settings;
+        }
+    }
+```
+
+Then, after the enclave image is loaded but before it is initialized, we add the additional memory pages containing the contents of the `oe_eeid_t` struct:
+
+```C
+        oe_sha256_context_t* hctx = &context->hash_context;
+        sgx_sigstruct_t* sigstruct = (sgx_sigstruct_t*)properties->sigstruct;
+        memcpy(eeid->sigstruct, (uint8_t*)sigstruct, sizeof(sgx_sigstruct_t));
+        oe_sha256_save(hctx, eeid->hash_state_H, eeid->hash_state_N);
+        eeid->data_vaddr = *vaddr;
+
+        uint64_t ee_sz = sizeof(oe_eeid_t) + eeid->data_size;
+        uint64_t epg_sz = eeid_pages_size(eeid);
+        uint64_t num_pages = epg_sz / OE_PAGE_SIZE;
+        assert(*vaddr == enclave_end - epg_sz);
+
+        oe_page_t* pages = (oe_page_t*)eeid;
+        if (ee_sz < epg_sz)
+        {
+            oe_page_t* tmp = (oe_page_t*)calloc(1, epg_sz);
+            memcpy(tmp, eeid, ee_sz);
+            pages = tmp;
+        }
+
+        OE_CHECK(_add_extra_data_pages(
+            context, enclave->addr, pages, num_pages, vaddr));
+
+        if (ee_sz < epg_sz)
+            free(pages);
+
+        OE_SHA256 ext_mrenclave;
+        oe_sha256_final(hctx, &ext_mrenclave);
+```
+
+Finally, since the changes invalidate the signature of the base image, we need to dynamically re-sign the enclave. For this, we use the `OE_DEBUG_SIGN_KEY`, which is well-known by all OE users. Note that applications should never rely on the apparent `MRSIGNER` of an EEID quote. Indeed, what we reflect to the application when we validate an EEID quote is the `MRSIGNER` of the base image.
+
+```C
```

Comment:
I think oesign does not need to be modified, but whoever verifies a quote needs to know the EEID to verify the quote from the extended enclave.

Reply:
I was wrong, I did end up having to modify oesign to extract the base-image hash state, vaddr, and entry_point to make those values available at image build time instead of enclave creation time. It's fairly straight-forward though.

Reply:
Could you include a brief paragraph on those changes as well here?

Reply:
Done!